Document guild time outs (#4075)
hammerandchisel/discord-api-docs: docs/resources/Guild.md

| premium_since?                | ?ISO8601 timestamp                              | when the user started [boosting](https://support.discord.com/hc/en-us/articles/360028038352-Server-Boosting-) the guild                  |
| deaf                          | boolean                                         | whether the user is deafened in voice channels                                                                                           |
| mute                          | boolean                                         | whether the user is muted in voice channels                                                                                              |
| pending?                      | boolean                                         | whether the user has not yet passed the guild's [Membership Screening](#DOCS_RESOURCES_GUILD/membership-screening-object) requirements   |
| permissions?                  | string                                          | total permissions of the member in the channel, including overwrites, returned when in the interaction object                            |

> | communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed. until then, they cannot interact with the server |

Reviewer: ```suggestion
| communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed; until then, they cannot interact with the server |
```

> | communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed; until then, they cannot interact with the server |

Reviewer 2: will do I guess :P

> | communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed; until then, they cannot interact with the server |

Reviewer 3: ```suggestion
| communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed; until then, they cannot interact with the server                                         |
```
table spacing

Reviewer 2: Done

> | communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed; until then, they cannot interact with the guild                                        |

Reviewer 4: should we reword the description to something more like:

"when the user's timeout will expire and the user's interaction in the guild is re-enabled. `null` if the user is not timed out."

`timeout / timed out / timing out` was avoided internally to keep the fields more generic and b/c `timeout` is an overloaded term, but it could be clearer to use `timeout / timed out` terminology at least in the descriptions here where we have more context, for the sake of having easily understandable docs. What do you think?

(same note for descriptions in the rest of the PR too)

Reviewer 2: Done

> | communication_disabled_until? | ?ISO8601 timestamp                              | till when the user is time-outed. This will leave the user completely unable to interact with the server till this time has been reached |

Reviewer 5: ```suggestion
| communication_disabled_until? | ?ISO8601 timestamp                              | timestamp of when the time out will be removed, leaving the user unable to interact with the server until then |
```

| premium_since?                | ?ISO8601 timestamp                              | when the user started [boosting](https://support.discord.com/hc/en-us/articles/360028038352-Server-Boosting-) the guild                |
| deaf                          | boolean                                         | whether the user is deafened in voice channels                                                                                         |
| mute                          | boolean                                         | whether the user is muted in voice channels                                                                                            |
| pending?                      | boolean                                         | whether the user has not yet passed the guild's [Membership Screening](#DOCS_RESOURCES_GUILD/membership-screening-object) requirements |
| permissions?                  | string                                          | total permissions of the member in the channel, including overwrites, returned when in the interaction object                          |

> | communication_disabled_until? | ?ISO8601 timestamp                              | when the user's timeout will expire and the user's interaction in the guild is re-enabled, null if the user is not timed out           |

Reviewer 6: Maybe we should refer to the FAQ page on Discord Support for how time outs work.
Since line `290` does this for how premium works, we could consider this as well:

https://support.discord.com/hc/en-us/articles/4413305239191-Time-Out-FAQ

Reviewer 4: this is a great idea 👍

> | Field                         | Type                                            | Description                                                                                                                                                                                                        |
> | ----------------------------- | ----------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
> | user?                         | [user](#DOCS_RESOURCES_USER/user-object) object | the user this guild member represents                                                                                                                                                                              |
> | nick?                         | ?string                                         | this users guild nickname                                                                                                                                                                                          |
> | avatar?                       | ?string                                         | the member's [guild avatar hash](#DOCS_REFERENCE/image-formatting)                                                                                                                                                 |
> | roles                         | array of snowflakes                             | array of [role](#DOCS_TOPICS_PERMISSIONS/role-object) object ids                                                                                                                                                   |
> | joined_at                     | ISO8601 timestamp                               | when the user joined the guild                                                                                                                                                                                     |
> | premium_since?                | ?ISO8601 timestamp                              | when the user started [boosting](https://support.discord.com/hc/en-us/articles/360028038352-Server-Boosting-) the guild                                                                                            |
> | deaf                          | boolean                                         | whether the user is deafened in voice channels                                                                                                                                                                     |
> | mute                          | boolean                                         | whether the user is muted in voice channels                                                                                                                                                                        |
> | pending?                      | boolean                                         | whether the user has not yet passed the guild's [Membership Screening](#DOCS_RESOURCES_GUILD/membership-screening-object) requirements                                                                             |
> | permissions?                  | string                                          | total permissions of the member in the channel, including overwrites, returned when in the interaction object                                                                                                      |
> | communication_disabled_until? | ?ISO8601 timestamp                              | when the user's [timeout](https://support.discord.com/hc/en-us/articles/4413305239191-Time-Out-FAQ) will expire and the user's interaction in the guild is re-enabled, null if the user is not timed out           |

Reviewer 3: ```suggestion
| Field                         | Type                                            | Description                                                                                                                                                                                                                          |
| ----------------------------- | ----------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| user?                         | [user](#DOCS_RESOURCES_USER/user-object) object | the user this guild member represents                                                                                                                                                                                                |
| nick?                         | ?string                                         | this users guild nickname                                                                                                                                                                                                            |
| avatar?                       | ?string                                         | the member's [guild avatar hash](#DOCS_REFERENCE/image-formatting)                                                                                                                                                                   |
| roles                         | array of snowflakes                             | array of [role](#DOCS_TOPICS_PERMISSIONS/role-object) object ids                                                                                                                                                                     |
| joined_at                     | ISO8601 timestamp                               | when the user joined the guild                                                                                                                                                                                                       |
| premium_since?                | ?ISO8601 timestamp                              | when the user started [boosting](https://support.discord.com/hc/en-us/articles/360028038352-Server-Boosting-) the guild                                                                                                              |
| deaf                          | boolean                                         | whether the user is deafened in voice channels                                                                                                                                                                                       |
| mute                          | boolean                                         | whether the user is muted in voice channels                                                                                                                                                                                          |
| pending?                      | boolean                                         | whether the user has not yet passed the guild's [Membership Screening](#DOCS_RESOURCES_GUILD/membership-screening-object) requirements                                                                                               |
| permissions?                  | string                                          | total permissions of the member in the channel, including overwrites, returned when in the interaction object                                                                                                                        |
| communication_disabled_until? | ?ISO8601 timestamp                              | when the user's [timeout](https://support.discord.com/hc/en-us/articles/4413305239191-Time-Out-FAQ) will expire and the user will be able to communicate in the guild again, null or a time in the past if the user is not timed out |
```

> | communication_disabled_until? | ?ISO8601 timestamp                              | when the user's [timeout](https://support.discord.com/hc/en-us/articles/4413305239191-Time-Out-FAQ) will expire and the user will be able to communicate in the guild again, null or a time in the past if the user is not timed out |

Reviewer 7: When can this property be missing?

Reviewer 2: It seems to be on every guild member object.
Had no case yet, where it was missing.

Reviewer 3: oh this is a good point - @hemu is it optional?

Reviewer 4: ah to be honest the optionality of fields from the perspective of the public api is always a tricky question to answer definitively for me. b/c the objects we define in the docs are defined more for convenience, so they can be referred to from other places in the docs rather than something that always maps 1:1 to how we use models internally. I think the best answer I can give is when we serialize the guild member model, we currently always seem to include the `communication_disabled_until` field. But this still doesn't 100% guarantee it is always included unless we thoroughly check every single endpoint where this model is included in some way currently. And this could change in the future at any time. But I *think* that means we can remove the optional?

Reviewer 4: since from a pure data model perspective `communication_disabled_until` isn't a core field for the member model like `joined_at` is, how about we leave it as optional just in case, since I'd imagine the opposite -- assuming it is required now and at some point in the future an endpoint whose document references this Guild Member Object in docs fails to include this field -- is a worse developer experience. What do you think?

Reviewer 2: We should leave it as optional :)

###### JSON Params

> | Field                        | Type                | Description                                                                                            | Permission       |
> | ---------------------------- | ------------------- | ------------------------------------------------------------------------------------------------------ | ---------------- |
> | nick                         | string              | value to set users nickname to                                                                         | MANAGE_NICKNAMES |
> | roles                        | array of snowflakes | array of role ids the member is assigned                                                               | MANAGE_ROLES     |
> | mute                         | boolean             | whether the user is muted in voice channels. Will throw a 400 if the user is not in a voice channel    | MUTE_MEMBERS     |
> | deaf                         | boolean             | whether the user is deafened in voice channels. Will throw a 400 if the user is not in a voice channel | DEAFEN_MEMBERS   |
> | channel_id                   | snowflake           | id of channel to move user to (if they are connected to voice)                                         | MOVE_MEMBERS     |
> | communication_disabled_until | ?ISO8601 timestamp  | how long the user can not interact with the guild; set to null to remove timeout                       | KICK_MEMBERS     |

Reviewer 3: api accepts end timestamps up to 28 days in the future

```suggestion
| Field                        | Type                | Description                                                                                                                | Permission       |
| ---------------------------- | ------------------- | -------------------------------------------------------------------------------------------------------------------------- | ---------------- |
| nick                         | string              | value to set users nickname to                                                                                             | MANAGE_NICKNAMES |
| roles                        | array of snowflakes | array of role ids the member is assigned                                                                                   | MANAGE_ROLES     |
| mute                         | boolean             | whether the user is muted in voice channels. Will throw a 400 if the user is not in a voice channel                        | MUTE_MEMBERS     |
| deaf                         | boolean             | whether the user is deafened in voice channels. Will throw a 400 if the user is not in a voice channel                     | DEAFEN_MEMBERS   |
| channel_id                   | snowflake           | id of channel to move user to (if they are connected to voice)                                                             | MOVE_MEMBERS     |
| communication_disabled_until | ?ISO8601 timestamp  | when the user should be able to interact with the guild again (up to 28 days in the future); set to null to remove timeout | KICK_MEMBERS     |
```

Reviewer 2: @Raiden-Ei You will get more informations on this soon enough.
This is not released.

Reviewer 3: > can I ask the source of that please

getting the error and then trying timestamps until i found the limit

###### JSON Params

| Field                        | Type                | Description                                                                                                                                                                                                                              | Permission       |
| ---------------------------- | ------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------- |

> | nick                         | string              | value to set users nickname to                                                                                                                                                                                                           | MANAGE_NICKNAMES |

Reviewer 3: ```suggestion
| nick                         | string              | value to set user's nickname to                                                                                                                                                                                                          | MANAGE_NICKNAMES |
```
should also be fixed in add guild member (related: prev comment)

> | Field                        | Type                | Description                                                                                                                  | Permission       |
> | ---------------------------- | ------------------- | ---------------------------------------------------------------------------------------------------------------------------- | ---------------- |
> | nick                         | string              | value to set users nickname to                                                                                               | MANAGE_NICKNAMES |
> | roles                        | array of snowflakes | array of role ids the member is assigned                                                                                     | MANAGE_ROLES     |
> | mute                         | boolean             | whether the user is muted in voice channels. Will throw a 400 if the user is not in a voice channel                          | MUTE_MEMBERS     |
> | deaf                         | boolean             | whether the user is deafened in voice channels. Will throw a 400 if the user is not in a voice channel                       | DEAFEN_MEMBERS   |
> | channel_id                   | snowflake           | id of channel to move user to (if they are connected to voice)                                                               | MOVE_MEMBERS     |
> | communication_disabled_until | ?ISO8601 timestamp  | when the user's timeout will expire and the user's interaction in the guild is re-enabled, null if the user is not timed out | MODERATE_MEMBERS |

Reviewer 3: ```suggestion
| Field                        | Type                | Description                                                                                                                                            | Permission       |
| ---------------------------- | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------- |
| nick                         | string              | value to set users nickname to                                                                                                                         | MANAGE_NICKNAMES |
| roles                        | array of snowflakes | array of role ids the member is assigned                                                                                                               | MANAGE_ROLES     |
| mute                         | boolean             | whether the user is muted in voice channels. Will throw a 400 if the user is not in a voice channel                                                    | MUTE_MEMBERS     |
| deaf                         | boolean             | whether the user is deafened in voice channels. Will throw a 400 if the user is not in a voice channel                                                 | DEAFEN_MEMBERS   |
| channel_id                   | snowflake           | id of channel to move user to (if they are connected to voice)                                                                                         | MOVE_MEMBERS     |
| communication_disabled_until | ?ISO8601 timestamp  | when the user's timeout will expire and the user's interaction in the guild is re-enabled (up to 28 days in the future), set to null to remove timeout | MODERATE_MEMBERS |
```

> | Field                        | Type                | Description                                                                                                                                            | Permission       |
> | ---------------------------- | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------- |
> | nick                         | string              | value to set users nickname to                                                                                                                         | MANAGE_NICKNAMES |
> | roles                        | array of snowflakes | array of role ids the member is assigned                                                                                                               | MANAGE_ROLES     |
> | mute                         | boolean             | whether the user is muted in voice channels. Will throw a 400 if the user is not in a voice channel                                                    | MUTE_MEMBERS     |
> | deaf                         | boolean             | whether the user is deafened in voice channels. Will throw a 400 if the user is not in a voice channel                                                 | DEAFEN_MEMBERS   |
> | channel_id                   | snowflake           | id of channel to move user to (if they are connected to voice)                                                                                         | MOVE_MEMBERS     |
> | communication_disabled_until | ?ISO8601 timestamp  | when the user's timeout will expire and the user's interaction in the guild is re-enabled (up to 28 days in the future), set to null to remove timeout | MODERATE_MEMBERS |

Reviewer 3: ```suggestion
| Field                        | Type                | Description                                                                                                                                                                                                                              | Permission       |
| ---------------------------- | ------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------- |
| nick                         | string              | value to set users nickname to                                                                                                                                                                                                           | MANAGE_NICKNAMES |
| roles                        | array of snowflakes | array of role ids the member is assigned                                                                                                                                                                                                 | MANAGE_ROLES     |
| mute                         | boolean             | whether the user is muted in voice channels. Will throw a 400 if the user is not in a voice channel                                                                                                                                      | MUTE_MEMBERS     |
| deaf                         | boolean             | whether the user is deafened in voice channels. Will throw a 400 if the user is not in a voice channel                                                                                                                                   | DEAFEN_MEMBERS   |
| channel_id                   | snowflake           | id of channel to move user to (if they are connected to voice)                                                                                                                                                                           | MOVE_MEMBERS     |
| communication_disabled_until | ?ISO8601 timestamp  | when the user's [timeout](https://support.discord.com/hc/en-us/articles/4413305239191-Time-Out-FAQ) will expire and the user will be able to communicate in the guild again (up to 28 days in the future), set to null to remove timeout | MODERATE_MEMBERS |
```
